BlueprintUI/Sources/Layout/GridRow.swift

```diff
@@ -0,0 +1,303 @@
+            items: [IndexedProportionalItem]
+        ) -> [IndexedSize] {
+            guard items.count > 0 else {
+                return []
+            }
+
+            let portionSum = items.map { $0.proportion }.reduce(0, +)
+            precondition(portionSum > 0, "Proportions of a GridRow must sum to a positive number. Found sum: \(portionSum).")
+            let scale = availableWidth / portionSum
+
+            return items.map { (index, proportion, content) in
+                let width = scale * proportion
+                let fixedWidthConstraint = SizeConstraint(width: .atMost(width), height: constraint.height)
+                let size = CGSize(width: width, height: content.measure(in: fixedWidthConstraint).height)
+                return (index, size)
+            }
+        }
+
+        /// When the layout width is undefined, size children relative to each other.
+        ///
+        /// Measure each child's width to find the maximum scale (points per portion). The maximum scale
+        /// can be applied to each child to provide widths which satisfy the proportional requirements
+        /// between children and provide enough space to layout each child.
+        ///
+        /// ## Example:
+        ///      - proportions:      1,     2,     3
+        ///      - measured widths:  10,    25,    24
+        ///      - scales:           10,    12.5,  8
+        ///      - max scale:        12.5
+        ///      - widths:           12.5,  25,    37.5
+        private func unconstrainedProportionalItemSizes(
+            in constraint: SizeConstraint,
+            items: [IndexedProportionalItem]
+        ) -> [IndexedSize] {
+            var scale: CGFloat = 0
+            items.forEach { (_, proportion, content) in
+                let width = content.measure(in: constraint).width
+                scale = max(scale, width / proportion)
+            }
+
+            return items.map { (index, proportion, content) in
+                let width = scale * proportion
+                let constraint = SizeConstraint(width: .atMost(width), height: constraint.height)
+                let size = CGSize(width: width, height: content.measure(in: constraint).height)
+                return (index, size)
+            }
+        }
+
+        private func _yVector(rowHeight: CGFloat) -> (CGFloat) -> (origin: CGFloat, height: CGFloat) {
+            switch verticalAlignment {
+            case .fill:
+                return { _ in (0, rowHeight) }
+            case .align(let id) where id == .top:
+                return { height in (0, height) }
+            case .align(let id) where id == .center:
+                return { height in ((rowHeight - height) / 2.0, height) }
+            case .align(let id) where id == .bottom:
+                return { height in (rowHeight - height,height) }
+            case .align:
+                fatalError("GridRow supports fill, top, center, and bottom alignment.")
```

Comment:
Hmm, I guess a follow-up for me would be to extract the cross-axis alignment logic from stacks so this is easier to implement.